File: test/harness-payload-matches-app.py
#!/usr/bin/env python3
"""THE PATCH SCRIPT'S PAYLOAD MUST STILL BE THE APP'S OWN TEXT.

`harness-v84-whatsnew.py` writes app-v84 by inserting text into app-v83's `index.html`. Every piece
of that text is a second copy of something that lives in the app, and a second copy drifts: the
hand-retyped version drifted FIVE times in one release, and after it was replaced with blocks lifted
verbatim it drifted again TWENTY MINUTES later, when the prefix fix moved one character in the app.

A boot check cannot see that. A stale-but-valid payload boots perfectly well, so `verify-rebuild.sh`
stayed green through it. Only a comparison catches it, and this is the comparison.

WHY THIS IS PYTHON AND NOT THE .mjs IT REPLACES. The first version matched `NAME = r\"\"\"...\"\"\"`
with a regex and checked FOUR payloads. An independent audit found it covered four of TWELVE
insertion sites: eight `cut()` calls carry their replacement text as inline literals, and it moved
three of them in the app while the check stayed green at 4/4. It also found that a payload written
`\"\"\"` instead of `r\"\"\"` was invisible to the regex -- and that form is already in the file. Both
gaps are the same shape: a check that can only see the spelling it was written against. Reading the
script's AST sees every string it actually writes, however it is spelled.

WHAT IT DOES NOT CLAIM. It does NOT require base + patch to equal HEAD. Fourteen commits of audit
fixes have no patch scripts, which README.md records. It requires only that what the script claims
to COPY is still a copy.
"""
import ast, sys, pathlib

ROOT = pathlib.Path(__file__).resolve().parent.parent
app = (ROOT / 'index.html').read_text()
script_path = ROOT / 'harness-v84-whatsnew.py'
tree = ast.parse(script_path.read_text())

# EXCLUDED BY DECISION, NOT BY ACCIDENT OF SPELLING -- which is exactly what the audit caught.
# BASE_HOOKS was invisible to the old regex because it is written `\"\"\"` rather than `r\"\"\"`, and
# excluding it happened to be right for a reason nobody had written down. It is the text as it
# exists in the BASE commit, which the script REMOVES; it is not meant to be in HEAD, and finding it
# there would mean the hook block never moved.
EXCLUDED = {
    'BASE_HOOKS': 'base-commit text that the script DELETES; it must NOT appear in HEAD',
}

def value_of(node, env):
    """Evaluate a string expression made of literals, names we know, and `+`."""
    if isinstance(node, ast.Constant) and isinstance(node.value, str):
        return node.value
    if isinstance(node, ast.Name) and node.id in env:
        return env[node.id]
    if isinstance(node, ast.BinOp) and isinstance(node.op, ast.Add):
        left, right = value_of(node.left, env), value_of(node.right, env)
        if left is None or right is None:
            return None
        return left + right
    return None

env, payloads = {}, []
for node in tree.body:
    if isinstance(node, ast.Assign) and len(node.targets) == 1 and isinstance(node.targets[0], ast.Name):
        v = value_of(node.value, env)
        if v is not None:
            name = node.targets[0].id
            env[name] = v
            if '\n' in v.strip():                      # a block, not a path or a flag
                payloads.append((name, v))

# Every cut(src, old, new, what) writes `new` INTO the app. That half is not self-checking the way
# the anchor is -- a stale `old` makes cut() die loudly, a stale `new` ships a file the app is not.
inserts = []
for node in ast.walk(tree):
    if isinstance(node, ast.Call) and isinstance(node.func, ast.Name) and node.func.id == 'cut' and len(node.args) >= 3:
        v = value_of(node.args[2], env)
        what = value_of(node.args[3], env) if len(node.args) >= 4 else '?'
        if v is not None and v.strip():
            inserts.append((what or '?', v))

ok = bad = 0
def check(label, text, expect_absent=False, reason=''):
    global ok, bad
    n = app.count(text)
    want = (n == 0) if expect_absent else (n == 1)
    if want:
        ok += 1
        print('  ok    %-58s %s' % (label, ('absent as intended' if expect_absent else '%d line(s), once' % len(text.split('\n')))))
    else:
        bad += 1
        print('  FAIL  %-58s appears %d time(s) in index.html' % (label, n))
        if reason:
            print('        %s' % reason)
        print('        re-extract it from index.html; do not hand-edit it back into agreement.')
        for line in text.split('\n'):
            if line.strip() and line not in app:
                print('        first line not in the app: %s' % line.strip()[:100])
                break

print('payload constants')
for name, text in payloads:
    if name in EXCLUDED:
        check('%s (excluded)' % name, text, expect_absent=True, reason=EXCLUDED[name])
    else:
        check(name, text)

print('\ntext written into the app by cut()')
for what, text in inserts:
    check(what, text)

# A check that finds nothing reports green loudest. Pin what it must have seen.
print('')
if len(payloads) < 4:
    print('  FAIL  only %d payload constant(s) found -- has the script been rewritten?' % len(payloads)); bad += 1
if len(inserts) < 8:
    print('  FAIL  only %d cut() insertion(s) found -- has the script been rewritten?' % len(inserts)); bad += 1
if bad == 0:
    print('  ok    saw %d payload constant(s) and %d cut() insertion(s)' % (len(payloads), len(inserts)))
    ok += 1

print('\n%d checks: %d passed, %d failed' % (ok + bad, ok, bad))
sys.exit(1 if bad else 0)

File: test/harness-payload-matches-app.mjs
// THE PATCH SCRIPT'S PAYLOAD MUST STILL BE THE APP'S OWN TEXT.
//
// harness-v84-whatsnew.py embeds several blocks of index.html verbatim, precisely because the
// hand-retyped versions drifted from the app FIVE times in one release. But "lifted verbatim" is a
// claim about the day it was lifted, not a property that holds: the app moved one character
// twenty minutes later -- `'chemowell-app'` gained its hyphen -- and the script went straight back
// to rebuilding a file the app no longer is, silently, with the boot gate still green because a
// stale-but-valid file boots perfectly well.
//
// So the guarantee gets a mechanism instead of a good intention. Every r""" block the script
// embeds must appear in index.html EXACTLY ONCE. If the app changes and the script is not
// re-extracted, this goes red the same day rather than at the next release.
//
// It deliberately does NOT require base + patch to equal HEAD -- fourteen commits of audit fixes
// have no patch scripts, which README.md records. It requires only that what the script DOES claim
// to copy is still a copy.
import { readFileSync } from 'node:fs';

const app = readFileSync('index.html', 'utf8');
const script = readFileSync('harness-v84-whatsnew.py', 'utf8');

// Pull every  NAME = r"""...."""  payload out of the patch script.
// The pattern allows anything between `=` and the `r"""` because MOVED_HOOKS is written as
// `MOVED_HOOKS = "\n" + r"""..."""`. The first version of this check anchored on `= r"""` and so
// silently covered three of the four payloads -- the hook block, which had ALREADY drifted once in
// this release (four hooks where the app exports five), was the one it could not see. A check that
// reports a clean board while blind to the block with the worst record is the exact defect class
// this file exists to catch, committed inside the catcher on its first run.
const blocks = [...script.matchAll(/^([A-Z_]+) = [^\n]*?r"""([\s\S]*?)"""/gm)]
  .map(([, name, body]) => ({ name, body }));

const EXPECTED = ['SURVIVORS', 'SNAPSHOT', 'BLOCK', 'MOVED_HOOKS'];
const missing = EXPECTED.filter((n) => !blocks.some((b) => b.name === n));
if (missing.length) {
  console.log(`  FAIL  these payloads were not matched at all: ${missing.join(', ')}`);
  console.log('        a payload this check cannot see is a payload nothing is checking.');
  process.exit(1);
}

if (blocks.length === 0) {
  console.log('  FAIL  no r""" payload blocks found in harness-v84-whatsnew.py -- has it been rewritten?');
  process.exit(1);
}

let pass = 0, fail = 0;
for (const { name, body } of blocks) {
  // MOVED_HOOKS carries the module's closing </script>, which is not part of the block in the app.
  const needle = body.replace(/\n?<\/script>\s*$/, '').trim();
  const n = app.split(needle).length - 1;
  if (n === 1) {
    pass++;
    console.log(`  ok    ${name} (${needle.split('\n').length} lines) appears once in index.html`);
  } else {
    fail++;
    console.log(`  FAIL  ${name} appears ${n} times in index.html -- the patch script has drifted`);
    console.log(`        re-extract it from index.html; do not hand-edit it back into agreement.`);
    const first = needle.split('\n').find((l) => l.trim() && !app.includes(l));
    if (first) console.log(`        first line not found in the app: ${first.trim().slice(0, 100)}`);
  }
}
console.log(`\n${pass + fail} checks: ${pass} passed, ${fail} failed`);
process.exit(fail ? 1 : 0);
